src/models/Level.php

```diff
@@ -647,18 +814,43 @@ private static function levelFromRow(Map<string, string> $row): Level {
   }
 
   // Get a single level.
-  public static async function gen(int $level_id): Awaitable<Level> {
-    $db = await self::genDb();
-
-    $result = await $db->queryf(
-      'SELECT * FROM levels WHERE id = %d LIMIT 1',
-      $level_id,
-    );
-
-    invariant($result->numRows() === 1, 'Expected exactly one result');
-    $level = self::levelFromRow($result->mapRows()[0]);
-
-    return $level;
+  /* HH_IGNORE_ERROR[4110]: Lines #827 and #835 prevent this function from failing to return */
+  public static async function gen(
+    int $level_id,
+    bool $refresh = false,
+  ): Awaitable<Level> {
+    $mc_result = self::getMCRecords('ALL_LEVELS');
+    if (!$mc_result || count($mc_result) === 0 || $refresh) {
+      $db = await self::genDb();
+      $all_levels = Map {};
+      $result = await $db->queryf('SELECT * FROM levels ORDER BY id');
+      foreach ($result->mapRows() as $row) {
+        $all_levels->add(
+          Pair {intval($row->get('id')), self::levelFromRow($row)},
+        );
+      }
+      self::setMCRecords('ALL_LEVELS', $all_levels);
+      invariant(
+        $all_levels->contains($level_id) !== false,
+        'level not found',
+      );
+      if ($all_levels->contains($level_id)) {
+        $level = $all_levels->get($level_id);
+        invariant($level instanceof Level, 'level should be of type Level');
+        return $level;
+      }
+    } else {
+      invariant(
+        $mc_result instanceof Map,
+        'cache return should be of type Map',
+      );
+      invariant($mc_result->contains($level_id) !== false, 'level not found');
+      if ($mc_result->contains($level_id)) {
```

Comment:
Since we have the `invariant`, let's remove the if statement and the HH_IGNOREs at the top. Same with the other places where we do this.